contracts/staking/StakeManager.sol

```diff
@@ -102,16 +118,26 @@ contract StakeManager is Validator, IStakeManager, RootChainable, Lockable {
     NFTCounter = NFTCounter.add(1);
   }
 
-  function unstake(uint256 validatorId) public onlyStaker(validatorId) {
-    require(validators[validatorId].activationEpoch > 0 && validators[validatorId].deactivationEpoch == 0);
+  function unstake(uint256 validatorId) external onlyStaker(validatorId) {
+    //Todo: add state here consider jail
+    require(validators[validatorId].activationEpoch > 0 &&
+      validators[validatorId].deactivationEpoch == 0 &&
+      validators[validatorId].status == Status.Active);
+
     uint256 amount = validators[validatorId].amount;
 
-    uint256 exitEpoch = currentEpoch.add(UNSTAKE_DELAY);
+    uint256 exitEpoch = currentEpoch.add(UNSTAKE_DELAY);// notice period
     validators[validatorId].deactivationEpoch = exitEpoch;
 
+    // unbond all delegatros in future
+    int256 delegationAmount = 0;
+    if (validators[validatorId].contractAddress != address(0x0)) {
+      delegationAmount = ValidatorContract(validators[validatorId].contractAddress).unBondAllLazy(exitEpoch);
+    }
+
     //  update future
     validatorState[exitEpoch].amount = (
-      validatorState[exitEpoch].amount - int256(amount));
+      validatorState[exitEpoch].amount - (int256(amount) + delegationAmount));
     validatorState[exitEpoch].stakerCount = (
       validatorState[exitEpoch].stakerCount - 1);
 
```

Comment:
Critical bug in `unstakeClaim`. You also need to check, validator had unstaked before claiming (i.e. `validators[validatorId].deactivationEpoch != 0`)

Reply:
ah! I see thanks.